syz-manager/repro.go

```diff
@@ -0,0 +1,157 @@
+	onlyOnce bool, mgr repoManagerView) *reproManager {
+	ret := &reproManager{
+		StatNumReproducing: stats.Create("reproducing", "Number of crashes being reproduced",
+			stats.Console, stats.NoGraph),
+		Done: make(chan *ReproResult, 10),
+
+		ctx:         ctx,
+		pool:        pool,
+		onlyOnce:    onlyOnce,
+		mgr:         mgr,
+		parallel:    make(chan struct{}, pool.Total()),
+		reproducing: map[string]bool{},
+		attempted:   map[string]bool{},
+	}
+	go ret.adjustPoolSize()
+	return ret
+}
+
+// startReproduction() is assumed to be called only once.
+// The agument is the maximum number of VMs dedicated to the bug reproduction.
+func (m *reproManager) startReproduction(VMs int) {
+	log.Logf(1, "starting reproductions (%d VMs available)", VMs)
+	for count := 1; count <= VMs; count++ {
+		need := m.needVMs(count)
+		if need > VMs {
+			break
+		}
+		m.parallel <- struct{}{}
+	}
+}
+
+func (m *reproManager) needVMs(repros int) int {
+	// For now, let's allocate 1.33 VMs per a reproducer thread.
+	return (repros*4 + 2) / 3
+}
+
+func (m *reproManager) canReproMore() bool {
+	select {
+	case <-m.parallel:
+		m.parallel <- struct{}{}
+		return true
+	default:
+		return false
+	}
+}
+
+func (m *reproManager) reproducingNow() map[string]bool {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	return maps.Clone(m.reproducing)
+}
+
+func (m *reproManager) handle(crash *Crash) {
+	m.mu.Lock()
+	// We don't want to reproduce the same bug twice at the same time.
+	if m.reproducing[crash.Title] {
+		m.mu.Unlock()
+		return
+	}
+	m.reproducing[crash.Title] = true
+	if m.onlyOnce && m.attempted[crash.Title] {
+		m.mu.Unlock()
+		// Try to reproduce each bug at most 1 time in this mode.
+		// Since we don't upload bugs/repros to dashboard, it likely won't have
+		// the reproducer even if we succeeded last time, and will repeatedly
+		// say it needs a repro.
+		return
+	}
+	m.attempted[crash.Title] = true
+	m.mu.Unlock()
+
+	defer func() {
+		m.mu.Lock()
+		m.reproducing[crash.Title] = false
+		m.mu.Unlock()
+	}()
+
+	log.Logf(1, "scheduled a reproduction of '%v'", crash.Title)
+
+	// Now wait for our turn to reproduce.
+	m.Pending.Add(1)
+	<-m.parallel
+	defer func() {
+		m.parallel <- struct{}{}
+	}()
+	m.Pending.Add(-1)
+
+	m.StatNumReproducing.Add(1)
+	defer m.StatNumReproducing.Add(-1)
+
+	log.Logf(1, "start reproducing '%v'", crash.Title)
+
+	res := m.mgr.runRepro(crash)
+
+	crepro := false
+	title := ""
+	if res.repro != nil {
+		crepro = res.repro.CRepro
+		title = res.repro.Report.Title
+	}
+	log.Logf(0, "repro finished '%v', repro=%v crepro=%v desc='%v' hub=%v from_dashboard=%v",
+		res.report0.Title, res.repro != nil, crepro, title, res.fromHub, res.fromDashboard,
+	)
+	m.Done <- res
+}
+
+func (m *reproManager) adjustPoolSize() {
+	for {
+		select {
+		case <-time.After(time.Minute):
+		case <-m.ctx.Done():
+			return
+		}
+		m.pool.ReserveForRun(m.needVMs(m.StatNumReproducing.Val()))
```

Comment:
Doing this with 1 minute delay for both increasing and reducing number of VMs in a separate goroutine looks clumsy. Can't we do it handle when we start/finish repros?

Reply:
If we do it on start/finish, we risk seeing a lot of "decrease by one, wait 1ms, increase by one". Adjusting it once per minute is more smooth.

If the `repro.go` code turns into something like repro queue instead of a set of parallel goroutines, it will probably be easier to avoid the jumps.

Reply:
A separate goroutine with minute sleeps does not look like the best solution for the jitter problem.

I am thinking about something along the following lines:

```
func newReproManager(VMs int, ...) *reproManager {
	ret := &reproManager{
		...
		parallel:    make(chan struct{}, reproCount(VMs)),
		start:       make(chan bool),
	}
}

func (m *reproManager) startReproduction() {
	close(m.start)
}

func (m *reproManager) handle(crash *Crash) {
	<-m.start

	m.mu.Lock()
	...
	m.reproducing[crash.Title] = true
	m.pool.ReserveForRun(m.needVMs(min(len(m.reproducing), cap(m.parallel)))
	m.mu.Unlock()

	defer func() {
		m.mu.Lock()
		delete(m.reproducing, crash.Title)
		m.pool.ReserveForRun(m.needVMs(min(len(m.reproducing), cap(m.parallel)))
		m.mu.Unlock()
	}()

	m.parallel <- struct{}{}
	defer func() {
		<-m.parallel
	}()

	...
}
```